File: backend/internal/usecase/smartsearch.go
package usecase

import (
	"context"
	"regexp"
	"strconv"
	"strings"

	"github.com/flapp/core/internal/domain"
)

// SmartSearchService turns a free-text request like "тёмные агрессивные 808 на
// 140 bpm" into a structured domain.SearchQuery: it recognises mood/genre words
// (mapped to tags), instrument words (mapped to the 40 categories) and tempo
// hints, in both Russian and English. Anything it cannot interpret falls back
// to a plain full-text search so the box is never a dead end.
type SmartSearchService struct {
	samples domain.SampleRepository
}

// NewSmartSearchService wires a smart-search service.
func NewSmartSearchService(samples domain.SampleRepository) *SmartSearchService {
	return &SmartSearchService{samples: samples}
}

// Interpretation reports what the parser understood, for UI feedback.
type Interpretation struct {
	Categories []domain.Category `json:"categories"`
	Tags       []string          `json:"tags"`
	MinBPM     int               `json:"minBpm"`
	MaxBPM     int               `json:"maxBpm"`
	FreeText   string            `json:"freeText"`
}

// SmartResult bundles the matches with the interpretation that produced them.
type SmartResult struct {
	Items          []*domain.Sample `json:"items"`
	Total          int              `json:"total"`
	Interpretation Interpretation   `json:"interpretation"`
}

// moodTerms maps Russian/English mood & genre words to canonical tags. These
// mirror the vocabulary produced by the TagGenerator so a search lines up with
// generated tags.
var moodTerms = map[string]string{
	"dark": "dark", "тёмные": "dark", "темные": "dark", "тёмный": "dark", "темный": "dark",
	"aggressive": "aggressive", "агрессивные": "aggressive", "агрессивный": "aggressive", "злые": "aggressive",
	"emotional": "emotional", "эмоциональные": "emotional", "эмоциональный": "emotional",
	"sad": "sad", "грустные": "sad", "грустный": "sad", "печальные": "sad",
	"melodic": "melodic", "мелодичные": "melodic", "мелодичный": "melodic",
	"trap": "trap", "трэп": "trap", "трап": "trap",
	"rage": "rage", "рейдж": "rage", "рэйдж": "rage",
	"drill": "drill", "дрилл": "drill",
	"cloud": "cloud", "клауд": "cloud",
	"jersey": "jersey", "джерси": "jersey",
	"pluggnb": "pluggnb", "плагг": "pluggnb", "плаг": "pluggnb",
	"futuristic": "futuristic", "футуристичные": "futuristic", "футуристичный": "futuristic",
	"hard": "hard", "жёсткие": "hard", "жесткие": "hard",
	"soft": "soft", "мягкие": "soft", "мягкий": "soft",
	"warm": "warm", "тёплые": "warm", "теплые": "warm",
	"bright": "bright", "яркие": "bright", "яркий": "bright",
	"lofi": "lofi", "лоуфай": "lofi", "лофай": "lofi",
}

// instrumentTerms maps words to one of the 11 categories.
var instrumentTerms = map[string]domain.Category{
	"808": domain.Cat808, "сабы": domain.Cat808, "саб": domain.Cat808,
	"kick": domain.CatKick, "кик": domain.CatKick, "бочка": domain.CatKick, "бочки": domain.CatKick,
	"snare": domain.CatSnare, "снейр": domain.CatSnare, "снэйр": domain.CatSnare, "малый": domain.CatSnare,
	"clap": domain.CatClap, "клэп": domain.CatClap, "клап": domain.CatClap, "хлопок": domain.CatClap,
	"hat": domain.CatHiHat, "hihat": domain.CatHiHat, "хэт": domain.CatHiHat, "хайхэт": domain.CatHiHat, "хай-хэт": domain.CatHiHat,
	"openhat": domain.CatOpenHat, "оупенхэт": domain.CatOpenHat,
	"crash": domain.CatOpenHat, "креш": domain.CatOpenHat,
	"ride": domain.CatOpenHat, "райд": domain.CatOpenHat,
	"cymbal": domain.CatOpenHat, "тарелка": domain.CatOpenHat, "тарелки": domain.CatOpenHat,
	"perc": domain.CatPerc, "перк": domain.CatPerc, "перкуссия": domain.CatPerc, "перкуссии": domain.CatPerc,
	"rim": domain.CatPerc, "римшот": domain.CatPerc,
	"tom": domain.CatPerc, "том": domain.CatPerc,
	"foley": domain.CatPerc, "фоли": domain.CatPerc,
	"vox": domain.CatVox, "vocal": domain.CatVox, "вокал": domain.CatVox, "вокалы": domain.CatVox,
	"chant": domain.CatVox, "чант": domain.CatVox,
	"fx": domain.CatFX, "эффект": domain.CatFX, "эффекты": domain.CatFX,
	"sweep": domain.CatFX, "свип": domain.CatFX,
	"impact": domain.CatFX, "импакт": domain.CatFX,
	"riser": domain.CatFX, "райзер": domain.CatFX,
	"midi": domain.CatFX, "миди": domain.CatFX,
	"texture": domain.CatFX, "текстура": domain.CatFX,
	"ambience": domain.CatFX, "эмбиенс": domain.CatFX,
	"piano": domain.CatLoop, "пиано": domain.CatLoop, "пианино": domain.CatLoop,
	"guitar": domain.CatLoop, "гитара": domain.CatLoop, "гитары": domain.CatLoop,
	"bell": domain.CatLoop, "белл": domain.CatLoop, "колокол": domain.CatLoop,
	"pluck": domain.CatLoop, "плак": domain.CatLoop,
	"synth": domain.CatLoop, "синт": domain.CatLoop, "синтезатор": domain.CatLoop,
	"pad": domain.CatLoop, "пэд": domain.CatLoop, "пад": domain.CatLoop,
	"bass": domain.CatLoop, "бас": domain.CatLoop, "басс": domain.CatLoop,
	"melody": domain.CatLoop, "мелодия": domain.CatLoop, "мелодии": domain.CatLoop,
	"loop": domain.CatLoop, "луп": domain.CatLoop, "лупы": domain.CatLoop,
	"drumloop": domain.CatDrumLoop, "драмлуп": domain.CatDrumLoop,
}

var bpmRe = regexp.MustCompile(`(?:(?:на|at)\s+)?(\d{2,3})\s*(?:bpm|бпм)`)
var bareNumRe = regexp.MustCompile(`\b(\d{2,3})\b`)

// Parse converts NL text into a SearchQuery and a description of what it found.
func (s *SmartSearchService) Parse(text string) (domain.SearchQuery, Interpretation) {
	lower := strings.ToLower(strings.TrimSpace(text))
	interp := Interpretation{}

	// Tempo: prefer an explicit "<n> bpm", else a bare 2–3 digit number.
	if m := bpmRe.FindStringSubmatch(lower); m != nil {
		if n, err := strconv.Atoi(m[1]); err == nil {
			interp.MinBPM, interp.MaxBPM = n-3, n+3
		}
		lower = bpmRe.ReplaceAllString(lower, " ")
	} else if m := bareNumRe.FindStringSubmatch(lower); m != nil {
		// Ignore "808" as a tempo — it is an instrument token.
		if m[1] != "808" {
			if n, err := strconv.Atoi(m[1]); err == nil && n >= 40 && n <= 220 {
				interp.MinBPM, interp.MaxBPM = n-3, n+3
			}
		}
	}
	if strings.Contains(lower, "медленны") || strings.Contains(lower, "slow") {
		interp.MaxBPM = 95
	}
	if strings.Contains(lower, "быстры") || strings.Contains(lower, "fast") {
		interp.MinBPM = 140
	}

	seenCat := map[domain.Category]bool{}
	seenTag := map[string]bool{}
	var leftover []string

	for _, tok := range tokenize(lower) {
		if cat, ok := instrumentTerms[tok]; ok {
			if !seenCat[cat] {
				seenCat[cat] = true
				interp.Categories = append(interp.Categories, cat)
			}
			continue
		}
		if tag, ok := moodTerms[tok]; ok {
			if !seenTag[tag] {
				seenTag[tag] = true
				interp.Tags = append(interp.Tags, tag)
			}
			continue
		}
		leftover = append(leftover, tok)
	}

	q := domain.SearchQuery{
		Categories: interp.Categories,
		Tags:       interp.Tags,
		MinBPM:     interp.MinBPM,
		MaxBPM:     interp.MaxBPM,
		Sort:       "used",
		Order:      "desc",
	}
	// Always pass unrecognised tokens to FTS so they narrow results further.
	// If nothing structured was recognised at all, fall back to the original
	// query string so partial CJK/emoji or uncommon names still match.
	if len(leftover) > 0 {
		q.Text = strings.Join(leftover, " ")
		interp.FreeText = q.Text
	} else if len(interp.Categories) == 0 && len(interp.Tags) == 0 && interp.MinBPM == 0 && interp.MaxBPM == 0 {
		q.Text = strings.TrimSpace(text)
		interp.FreeText = q.Text
	}
	return q, interp
}

// Search parses the text and runs the resulting query.
func (s *SmartSearchService) Search(ctx context.Context, text string, limit, offset int) (SmartResult, error) {
	q, interp := s.Parse(text)
	if limit > 0 {
		q.Limit = limit
	}
	q.Offset = offset
	items, total, err := s.samples.Search(ctx, q)
	if err != nil {
		return SmartResult{}, err
	}
	return SmartResult{Items: items, Total: total, Interpretation: interp}, nil
}

// tokenize splits text on non-alphanumeric runes, keeping Cyrillic and digits,
// and preserving the hyphen so "хай-хэт" survives as one token.
func tokenize(s string) []string {
	fields := strings.FieldsFunc(s, func(r rune) bool {
		if r == '-' {
			return false
		}
		switch {
		case r >= 'a' && r <= 'z':
			return false
		case r >= '0' && r <= '9':
			return false
		case r >= 'а' && r <= 'я' || r == 'ё':
			return false
		default:
			return true
		}
	})
	return fields
}

File: native/flapp-app/src/smartsearch.rs
// NL-поиск: «тёмные агрессивные 808 на 140 bpm» → структурный запрос
// (категории + теги + диапазон BPM), нераспознанное → свободный текст. Порт
// usecase/smartsearch.go (RU+EN словари настроений/инструментов/темпа).

use std::collections::HashMap;
use std::sync::OnceLock;

use regex::Regex;

use crate::classify::Category;
use Category::*;

/// Разобранный запрос: что удалось понять + свободный текст-остаток.
#[derive(Debug, Default, Clone, PartialEq)]
pub struct Parsed {
    pub categories: Vec<Category>,
    pub tags: Vec<String>,
    pub min_bpm: i32,
    pub max_bpm: i32,
    pub free_text: String,
}

fn instrument_terms() -> &'static HashMap<&'static str, Category> {
    static M: OnceLock<HashMap<&'static str, Category>> = OnceLock::new();
    M.get_or_init(|| {
        [
            ("808", C808), ("сабы", C808), ("саб", C808),
            ("kick", Kick), ("кик", Kick), ("бочка", Kick), ("бочки", Kick),
            ("snare", Snare), ("снейр", Snare), ("снэйр", Snare), ("малый", Snare),
            ("clap", Clap), ("клэп", Clap), ("клап", Clap), ("хлопок", Clap),
            ("hat", HiHat), ("hihat", HiHat), ("хэт", HiHat), ("хайхэт", HiHat), ("хай-хэт", HiHat),
            ("openhat", OpenHat), ("оупенхэт", OpenHat),
            ("crash", OpenHat), ("креш", OpenHat),
            ("ride", OpenHat), ("райд", OpenHat),
            ("cymbal", OpenHat), ("тарелка", OpenHat), ("тарелки", OpenHat),
            ("perc", Perc), ("перк", Perc), ("перкуссия", Perc), ("перкуссии", Perc),
            ("rim", Perc), ("римшот", Perc),
            ("tom", Perc), ("том", Perc),
            ("foley", Perc), ("фоли", Perc),
            ("vox", Vox), ("vocal", Vox), ("вокал", Vox), ("вокалы", Vox),
            ("chant", Vox), ("чант", Vox),
            ("fx", Fx), ("эффект", Fx), ("эффекты", Fx),
            ("sweep", Fx), ("свип", Fx),
            ("impact", Fx), ("импакт", Fx),
            ("riser", Fx), ("райзер", Fx),
            ("midi", Fx), ("миди", Fx),
            ("texture", Fx), ("текстура", Fx),
            ("ambience", Fx), ("эмбиенс", Fx),
            ("piano", Loop), ("пиано", Loop), ("пианино", Loop),
            ("guitar", Loop), ("гитара", Loop), ("гитары", Loop),
            ("bell", Loop), ("белл", Loop), ("колокол", Loop),
            ("pluck", Loop), ("плак", Loop),
            ("synth", Loop), ("синт", Loop), ("синтезатор", Loop),
            ("pad", Loop), ("пэд", Loop), ("пад", Loop),
            ("bass", Loop), ("бас", Loop), ("басс", Loop),
            ("melody", Loop), ("мелодия", Loop), ("мелодии", Loop),
            ("loop", Loop), ("луп", Loop), ("лупы", Loop),
            ("drumloop", DrumLoop), ("драмлуп", DrumLoop),
        ]
        .into_iter()
        .collect()
    })
}

fn mood_terms() -> &'static HashMap<&'static str, &'static str> {
    static M: OnceLock<HashMap<&'static str, &'static str>> = OnceLock::new();
    M.get_or_init(|| {
        [
            ("dark", "dark"), ("тёмные", "dark"), ("темные", "dark"), ("тёмный", "dark"), ("темный", "dark"),
            ("aggressive", "aggressive"), ("агрессивные", "aggressive"), ("агрессивный", "aggressive"), ("злые", "aggressive"),
            ("emotional", "emotional"), ("эмоциональные", "emotional"), ("эмоциональный", "emotional"),
            ("sad", "sad"), ("грустные", "sad"), ("грустный", "sad"), ("печальные", "sad"),
            ("melodic", "melodic"), ("мелодичные", "melodic"), ("мелодичный", "melodic"),
            ("trap", "trap"), ("трэп", "trap"), ("трап", "trap"),
            ("rage", "rage"), ("рейдж", "rage"), ("рэйдж", "rage"),
            ("drill", "drill"), ("дрилл", "drill"),
            ("cloud", "cloud"), ("клауд", "cloud"),
            ("jersey", "jersey"), ("джерси", "jersey"),
            ("pluggnb", "pluggnb"), ("плагг", "pluggnb"), ("плаг", "pluggnb"),
            ("futuristic", "futuristic"), ("футуристичные", "futuristic"), ("футуристичный", "futuristic"),
            ("hard", "hard"), ("жёсткие", "hard"), ("жесткие", "hard"),
            ("soft", "soft"), ("мягкие", "soft"), ("мягкий", "soft"),
            ("warm", "warm"), ("тёплые", "warm"), ("теплые", "warm"),
            ("bright", "bright"), ("яркие", "bright"), ("яркий", "bright"),
            ("lofi", "lofi"), ("лоуфай", "lofi"), ("лофай", "lofi"),
        ]
        .into_iter()
        .collect()
    })
}

fn bpm_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"(?:(?:на|at)\s+)?(\d{2,3})\s*(?:bpm|бпм)").unwrap())
}
fn bare_num_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"\b(\d{2,3})\b").unwrap())
}

/// Токенизация: разбить по не-алфанумерике, сохраняя a-z, 0-9, а-я, ё и дефис
/// (чтобы «хай-хэт» осталось одним токеном).
fn tokenize(s: &str) -> Vec<String> {
    let keep = |r: char| -> bool {
        r == '-' || r.is_ascii_lowercase() || r.is_ascii_digit() || ('а'..='я').contains(&r) || r == 'ё'
    };
    s.split(|c: char| !keep(c)).filter(|t| !t.is_empty()).map(|t| t.to_string()).collect()
}

/// Разбирает свободный текст в структурный запрос.
pub fn parse(text: &str) -> Parsed {
    let mut lower = text.trim().to_lowercase();
    let mut p = Parsed::default();

    // Темп: сначала явный «<n> bpm», иначе голое 2–3-значное число.
    if let Some(caps) = bpm_re().captures(&lower) {
        if let Ok(n) = caps[1].parse::<i32>() {
            p.min_bpm = n - 3;
            p.max_bpm = n + 3;
        }
        lower = bpm_re().replace_all(&lower, " ").into_owned();
    } else if let Some(caps) = bare_num_re().captures(&lower) {
        if &caps[1] != "808" {
            if let Ok(n) = caps[1].parse::<i32>() {
                if (40..=220).contains(&n) {
                    p.min_bpm = n - 3;
                    p.max_bpm = n + 3;
                }
            }
        }
    }
    if lower.contains("медленны") || lower.contains("slow") {
        p.max_bpm = 95;
    }
    if lower.contains("быстры") || lower.contains("fast") {
        p.min_bpm = 140;
    }

    let mut leftover: Vec<String> = Vec::new();
    for tok in tokenize(&lower) {
        if let Some(&cat) = instrument_terms().get(tok.as_str()) {
            if !p.categories.contains(&cat) {
                p.categories.push(cat);
            }
            continue;
        }
        if let Some(&tag) = mood_terms().get(tok.as_str()) {
            if !p.tags.iter().any(|t| t == tag) {
                p.tags.push(tag.to_string());
            }
            continue;
        }
        leftover.push(tok);
    }

    // Нераспознанные токены → в текст. Если структурного нет вообще — исходная
    // строка (частичные CJK/эмодзи/редкие имена всё равно найдутся).
    if !leftover.is_empty() {
        p.free_text = leftover.join(" ");
    } else if p.categories.is_empty() && p.tags.is_empty() && p.min_bpm == 0 && p.max_bpm == 0 {
        p.free_text = text.trim().to_string();
    }
    p
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn structured_ru_query() {
        let p = parse("тёмные агрессивные 808 на 140 bpm");
        assert_eq!(p.categories, vec![C808]);
        assert!(p.tags.iter().any(|t| t == "dark"));
        assert!(p.tags.iter().any(|t| t == "aggressive"));
        assert_eq!((p.min_bpm, p.max_bpm), (137, 143));
        assert_eq!(p.free_text, ""); // structured → no free text
    }

    #[test]
    fn instrument_to_loop_category() {
        let p = parse("dark melodic piano");
        assert_eq!(p.categories, vec![Loop]);
        assert!(p.tags.iter().any(|t| t == "dark"));
        assert!(p.tags.iter().any(|t| t == "melodic"));
    }

    #[test]
    fn free_text_fallback() {
        let p = parse("zzz weird name");
        assert!(!p.free_text.is_empty());
        assert!(p.categories.is_empty() && p.tags.is_empty());
    }

    #[test]
    fn tempo_words_and_808_not_tempo() {
        let slow = parse("slow 808");
        assert_eq!(slow.max_bpm, 95);
        assert_eq!(slow.categories, vec![C808]); // 808 is an instrument, not a tempo
    }
}
